### DISCUSSION TITLE
Load 1bpp PNG image and get the pixel data as byte array

### DISCUSSION THREAD

**Discussion by mrapavy** [SixLabors/ImageSharp#1405] [Q&A] [answered]
I have a 1bpp PNG image that I would like to load with ImageSharp, rotate it by 90 degrees and then get the pixel data of 1bpp image as byte array so it can be passed to another library which expects it like this.

This is where I am stuck:

```csharp
            byte[] DecodeImage(byte[] input1bppPng)
            {
                using (var image = Image.Load<???>(input1bppPng))
                {
                    image.Mutate(x => x.Rotate(RotateMode.Rotate90));
                    return MemoryMarshal.AsBytes(image.GetPixelSpan()).ToArray();
                }
            }
```

**Comment by JimBobSquarePants** [comment, marked as answer]
What's the actual memory layout of the buffer you require?

Pngs don't actually store the information as 1bpp. A png encoded with a bit depth of 1 means 1 bit per sample or per palette index (not per pixel). Which means the color palette of the indexed png has a maximum of 2 palette entries.

**Reply by antonfirsov** [reply]
I think the most likely case is that @mrapavy needs a bit array from a very compact PNG (not an expert on this but yeah, probably palettized I guess).

In this case what's needed is `Image.Load<L8>` to get the image data as `1 / 0` bytes, then a manual compression to a `byte[]` interpreted as a bit-array. We don't have any built-in helpers to achieve this second step.

@mrapavy does this help?

**Reply by JimBobSquarePants** [reply]
There's also `A8` which gaming engines use for 8 bit textures.

**Reply by mrapavy** [reply]
> I think the most likely case is that @mrapavy needs a bit array from a very compact PNG (not an expert on this but yeah, probably palettized I guess).
> 
> In this case what's needed is `Image.Load<L8>` to get the image data as `1 / 0` bytes, then a manual compression to a `byte[]` interpreted as a bit-array. We don't have any built-in helpers to achieve this second step.
> 
> @mrapavy does this help?

It does, thank you.

**Reply by mrapavy** [reply]
> What's the actual memory layout of the buffer you require?
> 
> Pngs don't actually store the information as 1bpp. A png encoded with a bit depth of 1 means 1 bit per sample or per palette index (not per pixel). Which means the color palette of the indexed png has a maximum of 2 palette entries.

The actual required memory layout is bit array (with palette of two colors). The comment posted by @antonfirsov helps.

**Reply by antonfirsov** [reply]
> In this case what's needed is `Image.Load<L8>` to get the image data as `1 / 0` bytes

@mrapavy this will be rather  `255 / 0` I guess
